contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,549 @@
+#include "upstream_request.h"
+
+#include <cstdint>
+#include <memory>
+#include <utility>
+
+#include "envoy/upstream/upstream.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/http/codes.h"
+#include "source/common/http/header_map_impl.h"
+#include "source/common/http/headers.h"
+#include "source/common/http/message_impl.h"
+#include "source/common/network/transport_socket_options_impl.h"
+#include "source/common/router/router.h"
+#include "source/extensions/common/proxy_protocol/proxy_protocol_header.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Upstreams {
+namespace Http {
+namespace Tcp {
+namespace Golang {
+
+FilterConfig::FilterConfig(
+    const envoy::extensions::upstreams::http::tcp::golang::v3alpha::Config proto_config, Dso::TcpUpstreamDsoPtr dso_lib)
+    : plugin_name_(proto_config.plugin_name()), so_id_(proto_config.library_id()),
+      so_path_(proto_config.library_path()), plugin_config_(proto_config.plugin_config()), dso_lib_(dso_lib){};
+
+FilterConfig::~FilterConfig() {
+  if (config_id_ > 0) {
+    dso_lib_->envoyGoTcpUpstreamDestroyPluginConfig(config_id_, 0);
+  }
+}
+
+void FilterConfig::newGoPluginConfig() {
+  ENVOY_LOG(debug, "tcp upstream initializing golang filter config");
+  std::string buf;
+  auto res = plugin_config_.SerializeToString(&buf);
+  ASSERT(res, "SerializeToString should always successful");
+  auto buf_ptr = reinterpret_cast<unsigned long long>(buf.data());
+  auto name_ptr = reinterpret_cast<unsigned long long>(plugin_name_.data());
+
+  config_ = new HttpConfigInternal(weak_from_this());
+  config_->plugin_name_ptr = name_ptr;
+  config_->plugin_name_len = plugin_name_.length();
+  config_->config_ptr = buf_ptr;
+  config_->config_len = buf.length();
+
+  config_id_ = dso_lib_->envoyGoOnTcpUpstreamConfig(config_);
+
+  if (config_id_ == 0) {
+    throw EnvoyException(
+        fmt::format("golang filter failed to parse plugin config: {} {}", so_id_, so_path_));
+  }
+
+  ENVOY_LOG(debug, "tcp upstream new plugin config, id: {}", config_id_);
+}
+
+Filter::Filter(FilterConfigSharedPtr config, const std::string cluster_name, const std::string route_name) : config_(config),
+  req_(new RequestInternal(*this)), encoding_state_(req_->encodingState()),
```

Comment:
RequestInternal could be inside Filter.

Reply:
then, envoyGoTcpUpstreamFinalize could be removed.

Reply:
done